tests/template_tests/test_base.py

```diff
@@ -1,8 +1,50 @@
 from django.template import Context, Template, Variable, VariableDoesNotExist
+from django.template.base import DebugLexer, Lexer, TokenType
 from django.test import SimpleTestCase
 from django.utils.translation import gettext_lazy
 
 
+class LexerTests(SimpleTestCase):
+    template_string = '''
+        <html><body>
+        {% if test %}<h1>{{ varvalue }}</h1>{% endif %}
+        </body></html>
+        {#comment {{not a var}} %{not a block}% #}
+        '''
+
+    expected_token_tuples = [
+        # (token_type, contents, lineno, position)
+        (TokenType.TEXT, '\n        <html><body>\n        ', 1, (0, 30)),
+        (TokenType.BLOCK, 'if test', 3, (30, 43)),
+        (TokenType.TEXT, '<h1>', 3, (43, 47)),
+        (TokenType.VAR, 'varvalue', 3, (47, 61)),
+        (TokenType.TEXT, '</h1>', 3, (61, 66)),
+        (TokenType.BLOCK, 'endif', 3, (66, 77)),
+        (TokenType.TEXT, '\n        </body></html>\n        ', 3, (77, 109)),
+        (TokenType.COMMENT, 'comment {{not a var}} %{not a block}%', 5, (109, 151)),
+        (TokenType.TEXT, '\n        ', 5, (151, 160)),
+    ]
+
+    def test_debuglexer(self):
+        """
+        Tests the DebugLexer's ability to tokenize a template string.
+        """
+
+        tokens = DebugLexer(self.template_string).tokenize()
+        token_tuples = [(t.token_type, t.contents, t.lineno, t.position) for t in tokens]
+        self.assertEqual(token_tuples, self.expected_token_tuples)
+
+    def test_lexer(self):
+        """
+        Tests the Lexer's ability to tokenize a template string.
+        """
+        tokens = Lexer(self.template_string).tokenize()
+        token_tuples = [(t.token_type, t.contents, t.lineno, t.position) for t in tokens]
+        # The non-debug lexer does not record position
+        none_position_token_tuples = [t[:-1] + (None,) for t in self.expected_token_tuples]
+        self.assertEqual(token_tuples, none_position_token_tuples)
```

Comment:
I would instead use the following pattern, sharing what you can:

```python
class LexerTestMixin:

    def test_tokenize(self):
        ...

class LexerTests(LexerTestMixin, SimpleTestCase):

    lexer_class = Lexer
    
    def make_expected(self):
        ...

class DebugLexerTests(LexerTestMixin, SimpleTestCase):

    lexer_class = DebugLexer

    def make_expected(self):
        ...
```

Reply:
@cjerdonek on this pattern, wouldn't `test_tokenize` want to use the `self.assertEqual` method of SimpleTestCase which it wouldn't have access to?

Reply:
Only `LexerTests` and `DebugLexerTests` would be calling `test_tokenize()`, which inherits from `SimpleTestCase`.